### PR TITLE
HDF5 interface for LHE-HDF5 event files

### PR DESCRIPTION
**Pull request by smrenna**
#### PR description:

This is a copy of LHESource to handle event generator files in the HDF5 format.
This uses the header-only HighFive library for handling the HDF5 files and information.
#### PR validation:

I have used the two tests in the LHEInterface tests directory:  testReader_cfg.py, which reads the hdf5 file and makes a root edm file and dumpLHE_cfg.py that prints out the stored information.

I can put some hdf5 files on lxplus for testing.

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by smrenna** [comment]
@bendavid might be interested in this

**Comment by smrenna** [comment]
I need some help here.    This is a direct copy of the structure of LHEReader, so why does one trigger a problem, but not the other?

Couldn't find the definition of iCatalog in CMSSW lxr.

**Comment by smrenna** [comment]
How do I properly remove the files flagged above?
e.g.:
GeneratorInterface/LHEInterface/src/HighFive/.gitmodules etc.

File: GeneratorInterface/LHEInterface/interface/lheh5.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <unistd.h>
+
+#include "GeneratorInterface/LHEInterface/src/HighFive/include/highfive/H5File.hpp"
+#include "GeneratorInterface/LHEInterface/src/HighFive/include/highfive/H5DataSet.hpp"
+using namespace HighFive;
```
Comment (davidlange6) [inline GeneratorInterface/LHEInterface/interface/lheh5.h:13]: using namespace is not allowed in headers.

File: GeneratorInterface/LHEInterface/src/HighFive/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+##Version 2.0 - 2018/07/19
```
Comment (davidlange6) [inline GeneratorInterface/LHEInterface/src/HighFive/CHANGELOG.md:1]: this is a copy of someone else's code
we show just make it an external package and let them maintain it.

**Comment by smrenna** [comment]
Agreed.   That was my hope.
How does one request a new external package?

Steve

**Comment by davidlange6** [comment]
Its done via a PR to cmsdist - the gen group experts can help

On Jun 22, 2020, at 3:04 PM, Stephen Mrenna <notifications@github.com<mailto:notifications@github.com>> wrote:


Agreed. That was my hope.
How does one request a new external package?

Steve

**Comment by agrohsje** [comment]
@SiewYan, @mkirsano is one of you available to help Steve preparing a PR to add HighFive as an external to https://github.com/cms-sw/cmsdist ?

**Comment by mkirsano** [comment]
So the proposal is to create a scram tool and add it to cmsdist?
What is this package, where it is described?
May be create a git issue with the necessary information?

**Comment by smrenna** [comment]
Hi,

Should I add the git issue to cms-sw/cmssw?

Steve

**Comment by smuzaffar** [comment]
@smrenna , yes please open an issue here https://github.com/cms-sw/cmsdist/issues . Provide the needed informstion, e.g. which external you would like to include in cmssw ( its web site). Any special information about the build configuration ( e.g. if it needs to be build with special flags etc.)

**Comment by silviodonato** [comment]
@smrenna what is the status of this PR?

**Comment by SiewYan** [comment]
@smrenna , i am not sure if this is relevant, I found the repo for HighFive: https://github.com/BlueBrain/HighFive, is this the external you needed?

**Comment by agrohsje** [comment]
Steve is currently away and should be back next week.

**Comment by smrenna** [comment]
Yes.   I apologize if I did not put this information into the original request.
Also, I originally requested an earlier tag, but I have since figured out how to use “master”.

Please ask if you have any additional questions.

Best,
Steve

**Comment by SiewYan** [comment]
Hello @smrenna , i am starting to prepare the tool file for HighFive, may i know which cmssw release you wanted it to be integrated?

**Comment by SiewYan** [comment]
@smrenna , the package is already integrated under 11_2_X [1] and released in CMSSW_11_2_0_pre2 [2]. Please let me know if you still need a backport to 10_6_X.

[1] https://github.com/cms-sw/cmsdist/pull/5996
[2] https://cmssdt.cern.ch/SDT/ReleaseNotes/CMSSW_11/CMSSW_11_2_0_pre2.html  (5996)

**Comment by mrodozov** [comment]
You cannot test 11_2 cmssw with 10_6 externals branch , highfive is already available in 11_2 . Pass the code checks, format and open a branch on 10_6 if you want to test highfive on 10_6 (cmsdist)

**Comment by SiewYan** [comment]
@mrodozov , the 10_6's PR is prepared here https://github.com/cms-sw/cmsdist/pull/6105 , @smrenna , can you close this PR in favor of 10_6's PR?

**Comment by mrodozov** [comment]
Ah now I understand. This is a HighFive itself in cmssw. It looks this PR contains the actual cmssw interface for 
HighFive
https://github.com/cms-sw/cmssw/pull/30312/files#diff-ab45c2eee17637adbbe74fbb65a820ff
@smrenna as the HighFive is now available as external package the cmssw code alone can remain.
And then open a PR in this repo for the interface (the client for HighFive) for branch https://github.com/cms-sw/cmssw/tree/CMSSW_10_6_X so we can test https://github.com/cms-sw/cmsdist/pull/6105 with it

**Comment by silviodonato** [comment]
https://github.com/cms-sw/cmsdist/pull/5996 has been merged right now

**Comment by silviodonato** [comment]
code-checks

**Comment by smrenna** [comment]
I updated the interface to be compatible with the master branch of the HighFive library, instead of a particular branch.

I ran the code-format tools too.

**Comment by smrenna** [comment]
Also, I was confused about how to pull from my gitlab repository into CMSSW, modify the code, and then push back.
As a result, I ended up doing an -ff-merge with one local branch to another, and, as a result, I picked up some baggage from other pull requests.   I am hoping this is not a problem.   Otherwise, please advise on what to do.

**Comment by kpedro88** [comment]
@smrenna this is indeed a problem, but it can be fixed.

http://cms-sw.github.io/tutorial-resolve-conflicts.html describes the correct procedures in significant detail. Short version:
* if you just need to recreate a working area, using the same CMSSW release as you originally used, the command is `git cms-checkout-topic`
* instead, if you need to move to a different (newer) CMSSW release, the command is `git cms-rebase-topic`
* the above can also be used to move to an older CMSSW release, but the `--old-base` argument is required for that to work properly

Recovering from an incorrect operation can be a bit more difficult. If you still have your local working areas, send me a message and I can try to help revert back to the previous state.

**Comment by bendavid** [comment]
Radical question:  Do we want/need to read future LHE files in CMSSW directly?  (As opposed to relying on the parser which will also have to exist in Pythia and probably Herwig)

**Comment by smrenna** [comment]
I didn’t think that was even negotiable?

What would it mean?    What about the creation of products, storing of provenance information, etc.?

Plus, we (Pythia) have been having disagreements on what that universal HDF5 parser would look like….

Steve

**Comment by bendavid** [comment]
This is the model which is already used for the SUSY parameter scan workflow (LHE's are parsed by pythia internal parser)

In this case all relevant information has to be forwarded through the HEPMCProduct and/or the GenRunInfo/GenLumiInfo/GenEventInfo product which may have generator-specific stuff as well.

So e.g. in that case the LHE weights get stored in the same place as the parton shower weights.

One probably needs to take extra care that the original unmodified LHE particles are really stored in the event record, clearly labelled and unmodified (which I'm pretty sure is not the case in pythia today)

**Comment by silviodonato** [comment]
please test

**Comment by mrodozov** [comment]
libs did not change , 
libGeneratorInterfaceBeamHaloGenerator needs hdf5 to get it's libs:
https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/BeamHaloGenerator/BuildFile.xml
for the other failure I think I'll have to add openmpi as requirement for hdf5 
Screenshot 2020-08-05 at 16 25 51

**Comment by smuzaffar** [comment]
please test with cms-sw/cmsdist#6152

**Comment by smrenna** [comment]
Don’t understand why this is being built.  It has nothing to do with the hdf5 reader.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by kpedro88** [comment]
`GeneratorInterface/Core` depends on `GeneratorInterface/LHEInterface`, which has modified header files in this PR.

The real question is why it can't find the libraries for hdf5_cpp and hdf5_hl_cpp even with the cmsdist PR provided...

**Comment by smuzaffar** [comment]
please test with cms-sw/cmsdist#6152
@kpedro88 , please see https://github.com/cms-sw/cmsdist/pull/6152#issuecomment-670172659

**Comment by smuzaffar** [comment]
please test with cms-sw/cmsdist#6152

**Comment by silviodonato** [comment]
After having merged https://github.com/cms-sw/cmsdist/pull/6160 the tests are ok. Do you have further comments @smrenna @cms-sw/generators-l2 ?

**Comment by SiewYan** [comment]
Hello, its looks all good to me. Unless @smrenna has other comments on this.

**Comment by silviodonato** [comment]
Kind reminder @cms-sw/generators-l2

**Comment by agrohsje** [comment]
Thanks @smrenna for the PR. As there was no additional feedback I am ok with the PR being merged.

**Comment by agrohsje** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**